sqldb/src/main.rs

```diff
@@ -0,0 +1,199 @@
+use std::collections::HashMap;
+use std::convert::Infallible;
+use std::sync::Arc;
+use tokio::sync::RwLock;
+use wasmbus_rpc::provider::prelude::MessageDispatch;
+use wasmbus_rpc::provider::prelude::*;
+use wasmcloud_interface_sqldb::{Column, ExecuteResult, FetchResult, Query, SqlDb, SqlDbReceiver};
+
+mod config;
+mod error;
+use error::DbError;
+
+mod types;
+
+// main (via provider_main) initializes the threaded tokio executor,
+// listens to lattice rpcs, handles actor links,
+// and returns only when it receives a shutdown message
+//
+fn main() -> Result<(), Box<dyn std::error::Error>> {
+    provider_main(SqlDbProvider::default())?;
+
+    eprintln!("sqldb provider exiting");
+    Ok(())
+}
+
+//pub(crate) type NTLS = native_tls::TlsConnector;
+pub(crate) type PgConnection = bb8_postgres::PostgresConnectionManager<NoTls>;
+pub(crate) type Pool = bb8_postgres::bb8::Pool<PgConnection>;
+
+/// sqldb capability provider implementation
+#[derive(Default, Clone, Provider)]
+#[services(SqlDb)]
+struct SqlDbProvider {
+    actors: Arc<RwLock<HashMap<String, Pool>>>,
+}
+
+/// use default implementations of provider message handlers
+impl ProviderDispatch for SqlDbProvider {}
+
+/// Handle connection pools for each link
+#[async_trait]
+impl ProviderHandler for SqlDbProvider {
+    /// Provider should perform any operations needed for a new link,
+    /// including setting up per-actor resources, and checking authorization.
+    /// If the link is allowed, return true, otherwise return false to deny the link.
+    async fn put_link(&self, ld: &LinkDefinition) -> RpcResult<bool> {
+        let config = config::load_config(ld)?;
+        let pool = config::create_pool(config).await?;
+        let mut update_map = self.actors.write().await;
+        update_map.insert(ld.actor_id.to_string(), pool);
+        Ok(true)
+    }
+
+    /// Handle notification that a link is dropped - close the connection
+    async fn delete_link(&self, actor_id: &str) {
+        let mut aw = self.actors.write().await;
+        if let Some(conn) = aw.remove(actor_id) {
+            // close all connections for this actor-link's pool
+            drop(conn);
+        }
+    }
+
+    /// Handle shutdown request by closing all connections
+    async fn shutdown(&self) -> Result<(), Infallible> {
+        let mut aw = self.actors.write().await;
+        // close all connections
+        for (_, conn) in aw.drain() {
+            drop(conn);
+        }
+        Ok(())
+    }
+}
+
+fn actor_id(ctx: &Context) -> Result<&String, RpcError> {
+    ctx.actor
+        .as_ref()
+        .ok_or_else(|| RpcError::InvalidParameter("no actor in request".into()))
+}
+
+/// SqlDb - SQL Database connections
+/// To use this capability, the actor must be linked
+/// with "wasmcloud:sqldb"
+/// wasmbus.contractId: wasmcloud:sqldb
+/// wasmbus.providerReceive
+#[async_trait]
+impl SqlDb for SqlDbProvider {
+    async fn execute(&self, ctx: &Context, query: &Query) -> RpcResult<ExecuteResult> {
```

Comment:
See my other comment in the `todo` example review - I think we should add an `execute` wrapper that takes a parameter map which then allows the provider to sanitize each field according to the requirements of the specific database being used, and removes the burden of producing safe/injection-proof SQL statements from the developers.

Reply:
prepared statements are also must-have feature. Not sure if they're in scope for first demo release.